From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wayne Chang <waynec@nvidia.com>
Date: Thu, 25 May 2023 08:05:01 +0000
Subject: [PATCH 086/240] NVIDIA: SAUCE: usb: typec: ucsi: Find connector
 fwnode without relying on order

The previous code relied on the order of "connector" nodes under the
ucsi device node to determine the correct fwnode handle for a connector.
However, the order of nodes should not be relied upon when parsing device
trees, especially when using live DT overlays.

This change modifies ucsi_find_fwnode() to find the correct connector
fwnode by reading the "reg" property of each connector node and
comparing it to the connector number. If only a single connector node
exists without a "reg" property, that single node is returned.

This ensures the correct connector fwnode is found regardless of node
order in the DT.

Bug 3960851
Bug 4063694

Change-Id: Id03dcece50cae369c4e1715ed8ba60cc6b28ee88
Signed-off-by: Wayne Chang <waynec@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-jammy/+/2918418
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2965133
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
Reviewed-by: Prathamesh Shete <pshete@nvidia.com>
Tested-by: Prathamesh Shete <pshete@nvidia.com>

diff --git a/drivers/usb/typec/ucsi/ucsi.c b/drivers/usb/typec/ucsi/ucsi.c
index b4b0c46d2fa44..dc828d057d8c5 100644
--- a/drivers/usb/typec/ucsi/ucsi.c
+++ b/drivers/usb/typec/ucsi/ucsi.c
@@ -1574,11 +1574,14 @@ static const struct typec_operations ucsi_ops = {
 static struct fwnode_handle *ucsi_find_fwnode(struct ucsi_connector *con)
 {
 	struct fwnode_handle *fwnode;
-	int i = 1;
+	u32 index;
+
+	device_for_each_child_node(con->ucsi->dev, fwnode) {
+		if (!fwnode_property_read_u32(fwnode, "reg", &index) && index+1 != con->num)
+			continue;
+		return fwnode;
+	}
 
-	device_for_each_child_node(con->ucsi->dev, fwnode)
-		if (i++ == con->num)
-			return fwnode;
 	return NULL;
 }
 
-- 
2.43.0

